From 39d901b3e64ec2b6e255e394eb0d7a1b29a1635c Mon Sep 17 00:00:00 2001
From: Kaan Alsan <alsan@chromium.org>
Date: Thu, 29 Jan 2026 11:49:10 -0800
Subject: [PATCH] Move drag offset data into DragSessionData

This is a refactoring with no changes in logic.

This moves the offset out of TabDragController and into DragSessionData.
This will allow clients (e.g. TabDragContext or TabDragTarget) to access
it and use it for layout calculations.

Change-Id: I12c48932651b806d5bc38987828f24a10aa67513
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/7532031
Commit-Queue: Kaan Alsan <alsan@chromium.org>
Reviewed-by: Alison Gale <agale@chromium.org>
Cr-Commit-Position: refs/heads/main@{#1576761}
---

--- a/chrome/browser/ui/views/tabs/dragging/drag_session_data.h
+++ b/chrome/browser/ui/views/tabs/dragging/drag_session_data.h
@@ -13,6 +13,7 @@
 #include "chrome/browser/ui/views/tabs/tab_slot_view.h"
 #include "components/tab_groups/tab_group_id.h"
 #include "components/tab_groups/tab_group_visual_data.h"
+#include "ui/gfx/geometry/vector2d_f.h"
 
 namespace content {
 class WebContents;
@@ -90,6 +91,10 @@ struct DragSessionData final {
   // user started dragging.
   size_t source_view_index_ = std::numeric_limits<size_t>::max();
 
+  // The offset of the mouse relative to the source dragged view's width and
+  // height.
+  gfx::Vector2dF mouse_offset_to_size_ratios;
+
   std::optional<tab_groups::TabGroupId> group() const {
     return group_drag_data_.has_value()
                ? std::make_optional(group_drag_data_.value().group)
--- a/chrome/browser/ui/views/tabs/dragging/dragging_tabs_session.cc
+++ b/chrome/browser/ui/views/tabs/dragging/dragging_tabs_session.cc
@@ -23,8 +23,7 @@ namespace {
 constexpr char kDragAmongTabsPresentationTimeHistogram[] =
     "Browser.TabDragging.DragAmongTabsPresentationTime";
 
-int CalculateMouseOffset(const DragSessionData& drag_data_,
-                         float offset_to_width_ratio_) {
+int CalculateMouseOffset(const DragSessionData& drag_data_) {
   std::vector<TabSlotView*> tabs_to_source(drag_data_.attached_views());
   TabSlotView* source_view = drag_data_.source_view_drag_data()->attached_view;
   tabs_to_source.erase(
@@ -32,7 +31,8 @@ int CalculateMouseOffset(const DragSessi
       tabs_to_source.end());
   const int new_x =
       TabStrip::GetSizeNeededForViews(tabs_to_source) - source_view->width() +
-      base::ClampRound(offset_to_width_ratio_ * source_view->width());
+      base::ClampRound(drag_data_.mouse_offset_to_size_ratios.x() *
+                       source_view->width());
 
   return new_x;
 }
@@ -43,13 +43,12 @@ DraggingTabsSession::DraggingTabsSession
     DragSessionData drag_data,
     TabDragContext& attached_context,
     TabDragPositioningDelegate& drag_position_delegate,
-    float offset_to_width_ratio_,
     bool initial_move,
     gfx::Point start_point_in_screen)
     : drag_data_(drag_data),
       attached_context_(attached_context),
       drag_position_delegate_(drag_position_delegate),
-      mouse_offset_(CalculateMouseOffset(drag_data_, offset_to_width_ratio_)),
+      mouse_offset_(CalculateMouseOffset(drag_data_)),
       initial_move_(initial_move),
       last_move_attached_context_loc_(
           views::View::ConvertPointFromScreen(&attached_context,
--- a/chrome/browser/ui/views/tabs/dragging/dragging_tabs_session.h
+++ b/chrome/browser/ui/views/tabs/dragging/dragging_tabs_session.h
@@ -27,7 +27,6 @@ class DraggingTabsSession final {
       DragSessionData drag_data,
       TabDragContext& attached_context,
       TabDragPositioningDelegate& drag_position_delegate,
-      float offset_to_width_ratio_,
       bool initial_move,
       gfx::Point point_in_screen);
   ~DraggingTabsSession();
--- a/chrome/browser/ui/views/tabs/dragging/dragging_tabs_session_browsertest.cc
+++ b/chrome/browser/ui/views/tabs/dragging/dragging_tabs_session_browsertest.cc
@@ -67,6 +67,7 @@ class DraggingTabsSessionBrowserTest : p
       drag_data.tab_drag_data_.back().attached_view = tab_view;
     }
     drag_data.source_view_index_ = source_index;
+    drag_data.mouse_offset_to_size_ratios.set_x(0.5);
 
     view_->GetDragContext()->StartedDragging(drag_data.attached_views());
 
@@ -96,7 +97,7 @@ IN_PROC_BROWSER_TEST_F(DraggingTabsSessi
   ASSERT_NE(drag_position_delegate, nullptr);
   const gfx::Point start_point = tab_0_view->GetBoundsInScreen().CenterPoint();
   DraggingTabsSession session(drag_data, *drag_context, *drag_position_delegate,
-                              0.5, true, start_point);
+                              true, start_point);
 
   // Swap them.
   const gfx::Point target_point =
--- a/chrome/browser/ui/views/tabs/dragging/tab_drag_controller.cc
+++ b/chrome/browser/ui/views/tabs/dragging/tab_drag_controller.cc
@@ -331,7 +331,6 @@ TabDragController::TabDragController()
       source_context_(nullptr),
       attached_context_(nullptr),
       can_release_capture_(true),
-      offset_to_width_ratio_(0),
       old_focused_view_tracker_(std::make_unique<views::ViewTracker>()),
       detach_behavior_(DetachBehavior::kDetachable),
       is_dragging_new_browser_(false),
@@ -477,14 +476,15 @@ TabDragController::Liveness TabDragContr
       ref->source_context_->GetWidget()->GetNativeWindow());
 
   if (source_view->width() > 0) {
-    ref->offset_to_width_ratio_ =
+    ref->drag_data_.mouse_offset_to_size_ratios.set_x(
         static_cast<float>(
             source_view->GetMirroredXInView(offset_from_source_view.x())) /
-        source_view->width();
+        source_view->width());
   }
   if (source_view->height() > 0) {
-    ref->offset_to_height_ratio_ =
-        static_cast<float>(offset_from_source_view.y()) / source_view->height();
+    ref->drag_data_.mouse_offset_to_size_ratios.set_y(
+        static_cast<float>(offset_from_source_view.y()) /
+        source_view->height());
   }
   ref->initial_selection_model_ = std::move(initial_selection_model);
 
@@ -2568,10 +2568,10 @@ gfx::Vector2d TabDragController::Calcula
   const views::View* source_view =
       drag_data_.attached_views()[drag_data_.source_view_index_];
   const gfx::Rect source_tab_bounds = source_view->bounds();
-  const int cursor_x_offset_within_tab =
-      base::ClampRound(source_tab_bounds.width() * offset_to_width_ratio_);
-  const int cursor_y_offset_within_tab =
-      base::ClampRound(source_tab_bounds.height() * offset_to_height_ratio_);
+  const int cursor_x_offset_within_tab = base::ClampRound(
+      source_tab_bounds.width() * drag_data_.mouse_offset_to_size_ratios.x());
+  const int cursor_y_offset_within_tab = base::ClampRound(
+      source_tab_bounds.height() * drag_data_.mouse_offset_to_size_ratios.y());
   gfx::Point desired_cursor_pos_in_widget(
       attached_context_->GetMirroredXInView(source_tab_bounds.x() +
                                             cursor_x_offset_within_tab),
@@ -2825,8 +2825,8 @@ void TabDragController::StartDraggingTab
 
   dragging_tabs_session_ = std::make_unique<DraggingTabsSession>(
       drag_data_, *attached_context_,
-      *attached_context_->GetPositioningDelegate(), offset_to_width_ratio_,
-      initial_move, start_point_in_screen);
+      *attached_context_->GetPositioningDelegate(), initial_move,
+      start_point_in_screen);
 }
 
 #if defined(USE_AURA)
--- a/chrome/browser/ui/views/tabs/dragging/tab_drag_controller.h
+++ b/chrome/browser/ui/views/tabs/dragging/tab_drag_controller.h
@@ -573,11 +573,6 @@ class TabDragController : public views::
   // DraggedTabView is constructed.
   gfx::Point start_point_in_screen_;
 
-  // Ratio of the x and y coordinates of the `source_view_offset` to the
-  // width and height of the source view.
-  float offset_to_width_ratio_ = 0;
-  float offset_to_height_ratio_ = 0;
-
   // Used to track the view that had focus in the window containing
   // `source_view_`. This is saved so that focus can be restored properly when
   // a drag begins and ends within this same window.
